draft-irtf-cfrg-opaque.md

```diff
@@ -830,20 +763,31 @@ Steps:
 2. N = Unblind(input.data_blind, Z)
 3. y = Finalize(PwdU, N, "RFCXXXX")
 4. RwdU = Harden(y, params)
-5. credentials_key = HKDF(salt=0, IKM=RwdU, "CredentialsKey", Nk)
-6. exporter_key = HKDF(salt=0, IKM=RwdU, "ExporterKey", Nk)
-7. n = random(Nn) # generate a fresh nonce randomly
-8. Credentials C with (skU=skU, pkS=response.pkS)
-9. (ct, t) = Seal(credentials_key, n, "", C)
-10. EnvU = concat(n, ct, t)
-11. Create RegistrationUpload upload with envelope value EnvU
-11. Output (upload, exporter_key)
+5. Create Credentials C according to server policy
+6. pt = SerializeExtensions(C.secret_credentials)
+7. nonce = random(Nn)
+8. pseudorandom_pad = HKDF-Expand(RwdU, contact(nonce, "Pad"), len(pt))
+9. auth_key = HKDF-Expand(RwdU, contact(nonce, "AuthKey"), Nk)
+10. exporter_key = HKDF-Expand(RwdU, concat(nonce, "ExporterKey"), Nk)
+11. ct = xor(pt, pseudorandom_pad)
+12. auth_data = SerializeExtensions(C.cleartext_credentials)
+13. t = HMAC(auth_key, concat(nonce, ct, concat(auth_data, aad)))
```

Comment:
Do we need aad? If so, where is it transmitted? It is not part of the envelope.

Reply:
Yes, I think so.